common/lib/xmodule/xmodule/tabs.py

```diff
@@ -174,62 +169,93 @@ def from_json(tab_dict):
-            'progress': ProgressTab,
-            'static_tab': StaticTab,
-            'peer_grading': PeerGradingTab,
-            'staff_grading': StaffGradingTab,
-            'open_ended': OpenEndedGradingTab,
-            'notes': NotesTab,
-            'edxnotes': EdxNotesTab,
-            'syllabus': SyllabusTab,
-            'instructor': InstructorTab,  # not persisted
-            'ccx_coach': CcxCoachTab,  # not persisted
-        }
-
-        tab_type = tab_dict.get('type')
-        if tab_type not in sub_class_types:
+        available_tab_types = CourseTabManager.get_tab_types()
+        tab_type_name = tab_dict.get('type')
+        if tab_type_name not in available_tab_types:
             raise InvalidTabsException(
-                'Unknown tab type {0}. Known types: {1}'.format(tab_type, sub_class_types)
+                'Unknown tab type {0}. Known types: {1}'.format(tab_type_name, available_tab_types)
             )
+        tab_type = available_tab_types[tab_dict['type']]
+        tab_type.validate(tab_dict)
+        if isinstance(tab_type, CourseFeatureTabType):
+            return tab_type.create_tab(tab_dict)
+        else:
+            return tab_type(tab_dict=tab_dict)
 
-        tab_class = sub_class_types[tab_dict['type']]
-        tab_class.validate(tab_dict)
-        return tab_class(tab_dict=tab_dict)
+
+class CourseTabManager(object):
+    """
+    A manager that handles the set of available course tabs.
+    """
+
+    @staticmethod
+    def get_tab_types():
+        """
+        Returns the list of available tab types.
+        """
+        if not hasattr(CourseTabManager, "_tab_types"):
+            tab_types = {
+                'courseware': CoursewareTab,
+                'course_info': CourseInfoTab,
+                'wiki': WikiTab,
+                'discussion': DiscussionTab,
+                'external_discussion': ExternalDiscussionTab,
+                'external_link': ExternalLinkTab,
+                'textbooks': TextbookTabs,
+                'pdf_textbooks': PDFTextbookTabs,
+                'html_textbooks': HtmlTextbookTabs,
+                'progress': ProgressTab,
+                'static_tab': StaticTab,
+                'peer_grading': PeerGradingTab,
+                'staff_grading': StaffGradingTab,
+                'open_ended': OpenEndedGradingTab,
+                'notes': NotesTab,
+                'syllabus': SyllabusTab,
+            }
+
+            # Add any course tabs that have been registered by features
+            # TODO: don't import openedx capabilities from common
+            from openedx.core.djangoapps.features.api import FeatureManager
+            for feature in FeatureManager.get_available_features().values():
+                if hasattr(feature, "course_tab"):
+                    tab_types[feature.name] = CourseFeatureTabType(feature)
+
+            CourseTabManager._tab_types = tab_types
+        return CourseTabManager._tab_types
```

Comment:
This class should probably just be a module-level function, rather than a `@staticmethod` on a class.